test/blackbox/common/network_namespace_framework.c

```diff
@@ -0,0 +1,468 @@
+        interface_handle = get_interface_handle_by_index(namespace_handle, if_no);
+
+      if(interface_handle->if_default_route_ip == NULL) {
+        peer_interface_handle = netns_get_priv_addr(test_state, interface_handle->fetch_ip_netns_name);
+        assert(peer_interface_handle);
+        interface_handle->if_default_route_ip  = ipv4_str_remove_cidr(peer_interface_handle->if_addr);
+      }
+    }
+  }
+}
+
+void netns_assign_ip_addresses(netns_state_t *test_state) {
+  int ns, j;
+  namespace_t *namespace_handle;
+  interface_t *interface_handle, *peer_interface_handle;
+  int sub_net;
+
+  if(test_state->public_net_addr == NULL) {
+    char *addr = malloc(INET6_ADDRSTRLEN);
+    assert(addr);
+    assert(strncpy(addr, DEFAULT_PUB_NET_ADDR, INET6_ADDRSTRLEN));
+    test_state->public_net_addr = addr;
+  }
+
+  for(ns = 0; ns < test_state->num_namespaces; ns++) {
+    namespace_handle = get_namespace_handle_by_index(test_state, ns);
+    assert(namespace_handle);
+    if(namespace_handle->type == BRIDGE) {
+      continue;
+    }
+    for(j = 0; j < namespace_handle->interfaces_no; j++) {
+      interface_handle = get_interface_handle_by_index(namespace_handle, j);
+      assert(interface_handle);
+
+      if(interface_handle->if_addr) {
+        continue;
+      }
+
+      // If fetch ip net namespace name is given get IP address from it, else get a public IP address
+
+      if(interface_handle->fetch_ip_netns_name) {
+        namespace_t *gw_netns_handle = find_namespace(test_state, interface_handle->fetch_ip_netns_name);
+        assert(gw_netns_handle);
+        assert(gw_netns_handle->static_config_net_addr);
+
+        increment_ipv4_cidr_str(gw_netns_handle->static_config_net_addr);
+        interface_handle->if_addr = strdup(gw_netns_handle->static_config_net_addr);
+      } else {
+        increment_ipv4_cidr_str(test_state->public_net_addr);
+        interface_handle->if_addr = strdup(test_state->public_net_addr);
+        if(interface_handle->if_default_route_ip == NULL && namespace_handle->type == HOST) {
+          interface_handle->if_default_route_ip = strdup(DEFAULT_GATEWAY_NET_ADDR);
+        }
+      }
+    }
+  }
+  netns_add_default_route_addr(test_state);
+}
+
+bool check_netns(const char *namespace_name) {
+  char cmd[200];
+  int sys_ret;
+
+  assert(snprintf(cmd, sizeof(cmd), "stat /var/run/netns/%s", namespace_name) >= 0);
+  sys_ret = system(cmd);
+
+  return !sys_ret;
+}
+
+void run_cmd_in_netns(char *namespace_name, char *cmd_str) {
+
+  assert(namespace_name && cmd_str);
+  assert(check_netns(namespace_name));
+
+   char *attach_argv[6] = { "ip", "netns", "exec", namespace_name, cmd_str, NULL };
+
+	// To daemonize, create a child process and detach it from its parent (this program)
+	if(fork() == 0) {
+		assert(daemon(1, 0) != -1);    // Detach from the parent process
+		assert(execv(attach_argv[0], attach_argv) != -1);   // Run exec() in the child process
```

Comment:
If you use `execv()` like this, then `cmd_str` can only point to an executable, but cannot contain any arguments. Maybe you should use `snprintf()+system()` like you do everywhere else, or change `attach_argv[]` to run `ip netns exec sh -c <cmd_str>`.